File: solutions_2453486_0/Python/kangz/tttt.py
import sys

def concat_str(args):
    s = ""
    for arg in args:
        s += str(arg)
    return s

def debug(*args):
    sys.stderr.write(concat_str(args) + "\n")

def printf(*args):
    debug(*args)
    print concat_str(args)

def int_input():
    return map(int, raw_input().split(' '))


#######################################################

def generate_lines(board):
    lines = []
    for i in range(4):
        lines.append([board[i][j] for j in range(4)])
        lines.append([board[j][i] for j in range(4)])
    lines.append([board[i][i] for i in range(4)])
    lines.append([board[3 - i][i] for i in range(4)])
    return lines


def read_input(        ):
    res = [[raw_input() for i in range(4)]]
    raw_input()
    return res

def solve(board):
    lines = generate_lines(board)
    Xwin = False
    Owin = False
    for l in lines:
        X = 0
        O = 0
        for c in l:
            if c == 'X' or c == 'T':
                X += 1
            if c == 'O' or c == 'T':
                O += 1
        if X == 4:
            Xwin = True
        if O == 4:
            Owin = True
    if Xwin:
        if Owin:
            return "Draw"
        else:
            return "X won"
    else:
        if Owin:
            return "O won"
        else:
            hasdot = any([any([board[i][j]=="." for j in range(4)]) for i in range(4)])
            if hasdot:
                return "Game has not completed"
            else:
                return "Draw"



#######################################################


for n_test_case in range(1, int(raw_input()) + 1):
    debug("Solving case ", n_test_case)

    printf("Case #", n_test_case, ": ", str(solve(*read_input())))

File: solutions_2453486_1/C++/bryanhe/tttt.cpp
#include <stdlib.h>
#include <stdio.h>
#include <iostream>
#include <algorithm>

using namespace std;

int main()
{
    int T;
    cin >> T;
    char board[4][4];

    for (int counter = 1; counter <= T; counter++) {

        for (int r = 0; r < 4; r++) {
            for (int c = 0; c < 4; c++) {
                cin >> board[r][c];
            }
        }

        bool XWin = false;
        bool OWin = false;

        for (int r = 0; r < 4; r++) {
            int X = 0;
            int O = 0;
            int T = 0;
            for (int c = 0; c < 4; c++) {
                if (board[r][c] == 'X') {
                    X++;
                }
                else if (board[r][c] == 'O') {
                    O++;
                }
                else if (board[r][c] == 'T') {
                    T++;
                }
            }
            if (X + T == 4) {
                XWin = true;
                break;
            }
            else if (O + T == 4) {
                OWin = true;
                break;
            }
        }

        if (!(XWin || OWin)) {
            for (int c = 0; c < 4; c++) {
                int X = 0;
                int O = 0;
                int T = 0;
                for (int r = 0; r < 4; r++) {
                    if (board[r][c] == 'X') {
                        X++;
                    }
                    else if (board[r][c] == 'O') {
                        O++;
                    }
                    else if (board[r][c] == 'T') {
                        T++;
                    }
                }
                if (X + T == 4) {
                    XWin = true;
                    break;
                }
                else if (O + T == 4) {
                    OWin = true;
                    break;
                }
            }

            if (!(XWin || OWin)) {
                int X = 0;
                int O = 0;
                int T = 0;
                for (int n = 0; n < 4; n++) {
                    if (board[n][n] == 'X') {
                        X++;
                    }
                    else if (board[n][n] == 'O') {
                        O++;
                    }
                    else if (board[n][n] == 'T') {
                        T++;
                    }
                }
                if (X + T == 4) {
                    XWin = true;
                }
                else if (O + T == 4) {
                    OWin = true;
                }


                X = 0;
                O = 0;
                T = 0;
                for (int n = 0; n < 4; n++) {
                    if (board[n][3 - n] == 'X') {
                        X++;
                    }
                    else if (board[n][3 - n] == 'O') {
                        O++;
                    }
                    else if (board[n][3 - n] == 'T') {
                        T++;
                    }
                }
                if (X + T == 4) {
                    XWin = true;
                }
                else if (O + T == 4) {
                    OWin = true;
                }
            }
        }

        int count = 0;
        for (int r = 0; r < 4; r++) {
            for (int c = 0; c < 4; c++) {
                if (board[r][c] != '.') {
                    count++;
                }
            }
        }

        cout << "Case #" << counter << ": ";

        if (XWin) {
            cout << "X won";
        }
        else if (OWin) {
            cout << "O won";
        }
        else if (count == 16) {
            cout << "Draw";
        }
        else {
            cout << "Game has not completed";
        }

        cout << "\n";

    }

    return 0;

}

